@rajbarik 

```
/Users/buildnode/jenkins/workspace/swift-PR-osx/branch-master/swift/test/SILOptimizer/sil_combine_protocol_conf.swift:247:11: error: CHECK: expected string not found in input
15:46:57 // CHECK: load [[S11]] : $*Builtin.Int64
15:46:57           ^
15:46:57 <stdin>:1376:51: note: scanning from here
15:46:57  %8 = struct_element_addr %7 : $*Int, #Int._value // user: %9
15:46:57                                                   ^
15:46:57 <stdin>:1376:51: note: with variable "S11" equal to "%8"
15:46:57  %8 = struct_element_addr %7 : $*Int, #Int._value // user: %9
15:46:57                                                   ^
15:46:57 <stdin>:1377:7: note: possible intended match here
15:46:57  %9 = load %8 : $*Builtin.Int32 // user: %20
15:46:57       ^
15:46:57 
15:46:57 --
15:46:57 
```

Looks like you need to tweak the regex for 32 bit. Please run `check-swift-validation-iphonesimulator-i386` locally.

--- a/include/swift/SILOptimizer/Analysis/ProtocolConformanceAnalysis.h
+++ b/include/swift/SILOptimizer/Analysis/ProtocolConformanceAnalysis.h
@@ -35,6 +35,8 @@ public:
   typedef SmallVector<NominalTypeDecl *, 8> NominalTypeList;
   typedef llvm::DenseMap<ProtocolDecl *, NominalTypeList>
       ProtocolConformanceMap;
+  typedef llvm::DenseMap<ProtocolDecl *, NominalTypeDecl *>
+      SoleConformingTypeMap;
 
   ProtocolConformanceAnalysis(SILModule *Mod)
       : SILAnalysis(SILAnalysisKind::ProtocolConformance), M(Mod) {
@@ -71,6 +73,10 @@ public:
                                              ConformsListIt->second.end())
                : ArrayRef<NominalTypeDecl *>();
   }
+  
+  /// Traverse ProtocolConformanceMapCache recursively to determine sole
+  /// conforming concrete type. 
+  NominalTypeDecl *findSoleConformingType(ProtocolDecl *Protocol);
 
 private:
   /// Compute inheritance properties.
@@ -79,8 +85,11 @@ private:
   /// The module.
   SILModule *M;
 
-  /// A cache that maps a protocol to its conformances
+  /// A cache that maps a protocol to its conformances.
   ProtocolConformanceMap ProtocolConformanceCache;
+
+  /// A cache that holds SoleConformingType for protocols.
+  SoleConformingTypeMap SoleConformingTypeCache;
 };
 
 } // namespace swift

--- a/include/swift/SILOptimizer/Utils/Existential.h
+++ b/include/swift/SILOptimizer/Utils/Existential.h
@@ -13,7 +13,10 @@
 #ifndef SWIFT_SILOPTIMIZER_UTILS_EXISTENTIAL_H
 #define SWIFT_SILOPTIMIZER_UTILS_EXISTENTIAL_H
 
+#include "swift/SIL/SILBuilder.h"
 #include "swift/SIL/SILInstruction.h"
+#include "swift/SILOptimizer/Analysis/ClassHierarchyAnalysis.h"
+#include "swift/SILOptimizer/Analysis/ProtocolConformanceAnalysis.h"
 
 namespace swift {
 
@@ -86,13 +89,24 @@ struct ConcreteExistentialInfo {
   // Constructs a valid ConcreteExistentialInfo object if successfull.
   ConcreteExistentialInfo(Operand &openedUse);
 
+  // This constructor initializes a ConcreteExistentialInfo based on already
+  // known ConcreteType and ProtocolDecl pair. It determines the
+  // OpenedArchetypeDef for the ArgOperand that will be used by unchecked_cast
+  // instructions to cast OpenedArchetypeDef to ConcreteType.
+  ConcreteExistentialInfo(Operand &ArgOperand, CanType ConcreteType,
+                          ProtocolDecl *Protocol);
+
   ConcreteExistentialInfo(ConcreteExistentialInfo &) = delete;
 
-  /// We do not not need to check for InitExistential not being not null
-  /// since if that were the case, we would have everything else null too.
+  /// For scenerios where ConcreteExistentialInfo is created using a known
+  /// ConcreteType and ProtocolDecl, both of InitExistential
+  /// and ConcreteValue can be null. So there is no need for explicit check for
+  /// not null for them instead we assert on (!InitExistential ||
+  /// ConcreteValue). 
   bool isValid() const {
+    assert(!InitExistential || ConcreteValue);
     return OpenedArchetype && OpenedArchetypeDef && ConcreteType &&
-           !ExistentialSubs.empty() && ConcreteValue;
+           !ExistentialSubs.empty();
   }
 
   // Do a conformance lookup on ConcreteType with the given requirement, P. If P

--- a/lib/SILOptimizer/Analysis/ProtocolConformanceAnalysis.cpp
+++ b/lib/SILOptimizer/Analysis/ProtocolConformanceAnalysis.cpp
@@ -67,4 +67,60 @@ void ProtocolConformanceAnalysis::init() {
   }
 }
 
+/// Recursively traverse the conformance lists to determine sole conforming
+/// class, struct or enum type.
+NominalTypeDecl *
+ProtocolConformanceAnalysis::findSoleConformingType(ProtocolDecl *Protocol) {
+
+  /// First check in the SoleConformingTypeCache.
+  auto SoleConformingTypeIt = SoleConformingTypeCache.find(Protocol);
+  if (SoleConformingTypeIt != SoleConformingTypeCache.end())
+    return SoleConformingTypeIt->second;
+
+  SmallVector<ProtocolDecl *, 8> PDWorkList;
+  SmallPtrSet<ProtocolDecl *, 8> VisitedPDs;
+  NominalTypeDecl *SoleConformingNTD = nullptr;
+  PDWorkList.push_back(Protocol);
+  while (!PDWorkList.empty()) {
+    auto *PD = PDWorkList.pop_back_val();
+    // Protocols must have internal or lower access.
+    if (PD->getEffectiveAccess() > AccessLevel::Internal) {
+      return nullptr;
+    }
+    VisitedPDs.insert(PD);
+    auto NTDList = getConformances(PD);
+    for (auto *ConformingNTD : NTDList) {
+      // Recurse on protocol types.
+      if (auto *Proto = dyn_cast<ProtocolDecl>(ConformingNTD)) {
+        // Ignore visited protocol decls.
+        if (!VisitedPDs.count(Proto))
+          PDWorkList.push_back(Proto);
+      } else { // Classes, Structs and Enums are added here.
+        // Bail if more than one conforming types were found.
+        if (SoleConformingNTD && ConformingNTD != SoleConformingNTD) {
+          return nullptr;
+        } else {
+          SoleConformingNTD = ConformingNTD;
+        }
+      }
+    }
+  }
+
+  // Bail if we did not find a sole conforming type.
+  if (!SoleConformingNTD)
+    return nullptr;
+
+  // Generic declarations are ignored.
+  if (SoleConformingNTD->isGenericContext()) {
+    return nullptr;
+  }
+
+  // Populate SoleConformingTypeCache.
+  SoleConformingTypeCache.insert(std::pair<ProtocolDecl *, NominalTypeDecl *>(
+      Protocol, SoleConformingNTD));
+
+  // Return SoleConformingNTD.
+  return SoleConformingNTD;
+}
+
 ProtocolConformanceAnalysis::~ProtocolConformanceAnalysis() {}

--- a/lib/SILOptimizer/SILCombiner/SILCombine.cpp
+++ b/lib/SILOptimizer/SILCombiner/SILCombine.cpp
@@ -347,12 +347,15 @@ class SILCombine : public SILFunctionTransform {
   void run() override {
     auto *AA = PM->getAnalysis<AliasAnalysis>();
     auto *DA = PM->getAnalysis<DominanceAnalysis>();
+    auto *PCA = PM->getAnalysis<ProtocolConformanceAnalysis>();
+    auto *CHA = PM->getAnalysis<ClassHierarchyAnalysis>();
 
     SILOptFunctionBuilder FuncBuilder(*this);
     // Create a SILBuilder with a tracking list for newly added
     // instructions, which we will periodically move to our worklist.
     SILBuilder B(*getFunction(), &TrackingList);
-    SILCombiner Combiner(FuncBuilder, B, AA, DA, getOptions().RemoveRuntimeAsserts);
+    SILCombiner Combiner(FuncBuilder, B, AA, DA, PCA, CHA,
+                         getOptions().RemoveRuntimeAsserts);
     bool Changed = Combiner.runOnFunction(*getFunction());
     assert(TrackingList.empty() &&
            "TrackingList should be fully processed by SILCombiner");

--- a/lib/SILOptimizer/SILCombiner/SILCombiner.h
+++ b/lib/SILOptimizer/SILCombiner/SILCombiner.h
@@ -25,6 +25,8 @@
 #include "swift/SIL/SILInstruction.h"
 #include "swift/SIL/SILValue.h"
 #include "swift/SIL/SILVisitor.h"
+#include "swift/SILOptimizer/Analysis/ClassHierarchyAnalysis.h"
+#include "swift/SILOptimizer/Analysis/ProtocolConformanceAnalysis.h"
 #include "swift/SILOptimizer/Utils/CastOptimizer.h"
 #include "swift/SILOptimizer/Utils/Existential.h"
 #include "swift/SILOptimizer/Utils/Local.h"
@@ -118,6 +120,14 @@ class SILCombiner :
 
   DominanceAnalysis *DA;
 
+  // Determine the set of types a protocol conforms to in whole-module
+  // compilation mode.
+  ProtocolConformanceAnalysis *PCA;
+
+  // Class hierarchy analysis needed to confirm no derived classes of a sole
+  // conforming class.
+  ClassHierarchyAnalysis *CHA;
+
   /// Worklist containing all of the instructions primed for simplification.
   SILCombineWorklist Worklist;
 
@@ -139,8 +149,9 @@ class SILCombiner :
 public:
   SILCombiner(SILOptFunctionBuilder &FuncBuilder,
               SILBuilder &B, AliasAnalysis *AA, DominanceAnalysis *DA,
+              ProtocolConformanceAnalysis *PCA, ClassHierarchyAnalysis *CHA,
               bool removeCondFails)
-      : AA(AA), DA(DA), Worklist(), MadeChange(false),
+      : AA(AA), DA(DA), PCA(PCA), CHA(CHA), Worklist(), MadeChange(false),
         RemoveCondFails(removeCondFails), Iteration(0), Builder(B),
         CastOpt(FuncBuilder,
                 /* ReplaceInstUsesAction */
@@ -288,17 +299,21 @@ private:
                                               const ConcreteExistentialInfo &CEI,
                                               SILBuilderContext &BuilderCtx);
 
+  // Common utility function to replace the WitnessMethodInst using a
+  // BuilderCtx.
+  void replaceWitnessMethodInst(WitnessMethodInst *WMI,
+                                SILBuilderContext &BuilderCtx,
+                                CanType ConcreteType,
+                                const ProtocolConformanceRef ConformanceRef);
+
   SILInstruction *
   propagateConcreteTypeOfInitExistential(FullApplySite Apply,
                                          WitnessMethodInst *WMI);
   SILInstruction *propagateConcreteTypeOfInitExistential(FullApplySite Apply);
 
-  // Common utility function to replace the WitnessMethodInst using a
-  // BuilderCtx.
-  void replaceWitnessMethodInst(FullApplySite Apply, WitnessMethodInst *WMI,
-                                SILBuilderContext &BuilderCtx,
-                                const CanType &ConcreteType,
-                                const ProtocolConformanceRef ConformanceRef);
+  /// Propagate concrete types from ProtocolConformanceAnalysis.
+  SILInstruction *propagateSoleConformingType(FullApplySite Apply,
+                                              WitnessMethodInst *WMI);
 
   /// Perform one SILCombine iteration.
   bool doOneIteration(SILFunction &F, unsigned Iteration);

--- a/lib/SILOptimizer/SILCombiner/SILCombinerApplyVisitors.cpp
+++ b/lib/SILOptimizer/SILCombiner/SILCombinerApplyVisitors.cpp
@@ -605,19 +605,98 @@ SILCombiner::optimizeConcatenationOfStringLiterals(ApplyInst *AI) {
 
 /// This routine replaces the old witness method inst with a new one.
 void SILCombiner::replaceWitnessMethodInst(
-    FullApplySite Apply, WitnessMethodInst *WMI, SILBuilderContext &BuilderCtx,
-    const CanType &ConcreteType, const ProtocolConformanceRef ConformanceRef) {
+    WitnessMethodInst *WMI, SILBuilderContext &BuilderCtx, CanType ConcreteType,
+    const ProtocolConformanceRef ConformanceRef) {
   SILBuilderWithScope WMIBuilder(WMI, BuilderCtx);
   auto *NewWMI = WMIBuilder.createWitnessMethod(
       WMI->getLoc(), ConcreteType, ConformanceRef, WMI->getMember(),
       WMI->getType());
-  MutableArrayRef<Operand> Operands = Apply.getInstruction()->getAllOperands();
-  for (auto &Op : Operands) {
-    if (Op.get() == WMI)
-      Op.set(NewWMI);
-  }
+  WMI->replaceAllUsesWith(NewWMI);
   if (WMI->use_empty())
     eraseInstFromFunction(*WMI);
+}
+
+//  This function propagates concrete type of existential self argument using
+//  ProtocolConformanceAnalysis. The concrete type of self can be a class,
+//  struct, or an enum. It replaces the witness_method instruction
+//  with one that has a concrete type, allowing other optimizations to
+//  devirtualize it later.
+SILInstruction *
+SILCombiner::propagateSoleConformingType(FullApplySite Apply,
+                                         WitnessMethodInst *WMI) {
+  // If WMI has concrete conformance, it can be optimized.
+  if (WMI->getConformance().isConcrete())
+    return nullptr;
+
+  // If the lookup type is an opened existential type,
+  // it cannot be made concrete.
+  if (!WMI->getLookupType()->isOpenedExistential())
+    return nullptr;
+
+  // If the witness method mutates self, we cannot replace self.
+  if (Apply.getOrigCalleeType()->getSelfParameter().isIndirectMutating())
+    return nullptr;
+
+  // Only applicable in whole-module compilation.
+  if (!Apply.getModule().isWholeModule())
+    return nullptr;
+
+  auto *PD = WMI->getLookupProtocol();
+
+  // Determine the sole conforming type.
+  auto *NTD = PCA->findSoleConformingType(PD);
+  if (!NTD)
+    return nullptr;
+
+  // Sole conforming class should not be open access or have any derived class.
+  ClassDecl *CD;
+  if ((CD = dyn_cast<ClassDecl>(NTD)) &&
+      (CD->getEffectiveAccess() == AccessLevel::Open ||
+       CHA->hasKnownDirectSubclasses(CD))) {
+    return nullptr;
+  }
+
+  // Create SIL type for the concrete type.
+  auto ElementType = NTD->getDeclaredType();
+  auto ConcreteType = ElementType->getCanonicalType();
+  auto &M = Builder.getModule();
+
+  /// Determine OpenedArchetypeDef and SubstituionMap.
+  ConcreteExistentialInfo CEI(Apply.getSelfArgumentOperand(), ConcreteType, PD);
+  if (!CEI.isValid())
+    return nullptr;
+
+  if (!CEI.InitExistential) {
+    // Create SIL type for the concrete type.
+    SILType ConcreteSILType = M.Types.getLoweredType(ConcreteType);
+
+    // Prepare the code by adding UncheckedCast instructions that cast opened
+    // existentials to concrete types. Set the ConcreteValue of CEI.
+    if (auto *OER = dyn_cast<OpenExistentialRefInst>(CEI.OpenedArchetypeDef)) {
+      auto *URCI =
+          Builder.createUncheckedRefCast(OER->getLoc(), OER, ConcreteSILType);
+      CEI.ConcreteValue = URCI;
+    } else if (auto *OEA =
+                   dyn_cast<OpenExistentialAddrInst>(CEI.OpenedArchetypeDef)) {
+      auto *UACI = Builder.createUncheckedAddrCast(
+          OEA->getLoc(), OEA, ConcreteSILType.getAddressType());
+      CEI.ConcreteValue = UACI;
+    } else {
+      llvm_unreachable(
+          "Unhandled Argument Type in propagateSoleConformingType");
+    }
+  }
+
+  assert(CEI.ConcreteValue);
+
+  /// Replace the old WitnessMethod with a new one that has concrete type and
+  /// conformance.
+  SILBuilderContext BuilderCtx(M, Builder.getTrackingList());
+  replaceWitnessMethodInst(WMI, BuilderCtx, ConcreteType,
+                           *(CEI.ExistentialSubs.getConformances().begin()));
+  /// Create the new apply instruction using the concrete type.
+  auto *NewAI = createApplyWithConcreteType(Apply, CEI, BuilderCtx);
+  return NewAI;
 }
 
 /// Given an Apply and an argument value produced by InitExistentialAddrInst,
@@ -799,7 +878,6 @@ SILCombiner::propagateConcreteTypeOfInitExistential(FullApplySite Apply,
   if (WMI->getConformance().isConcrete())
     return nullptr;
 
-
   // If the lookup type is not an opened existential type,
   // it cannot be made more concrete.
   if (!WMI->getLookupType()->isOpenedExistential())
@@ -840,7 +918,7 @@ SILCombiner::propagateConcreteTypeOfInitExistential(FullApplySite Apply,
   // with it.
   if (CEI.ConcreteType != WMI->getLookupType() ||
       SelfConformance != WMI->getConformance()) {
-    replaceWitnessMethodInst(Apply, WMI, BuilderCtx, CEI.ConcreteType,
+    replaceWitnessMethodInst(WMI, BuilderCtx, CEI.ConcreteType,
                              SelfConformance);
   }
   // Try to rewrite the apply.
@@ -1101,7 +1179,12 @@ SILInstruction *SILCombiner::visitApplyInst(ApplyInst *AI) {
   // (apply (witness_method)) -> propagate information about
   // a concrete type from init_existential_addr or init_existential_ref.
   if (auto *WMI = dyn_cast<WitnessMethodInst>(AI->getCallee())) {
-    propagateConcreteTypeOfInitExistential(AI, WMI);
+    if (!propagateConcreteTypeOfInitExistential(AI, WMI)) {
+      if (auto *WitnessMethod = dyn_cast<WitnessMethodInst>(AI->getCallee())) {
+        // Propagate concrete type from ProtocolConformanceAnalysis.
+        propagateSoleConformingType(AI, WitnessMethod);
+      }
+    }
     return nullptr;
   }
 
@@ -1219,10 +1302,16 @@ SILInstruction *SILCombiner::visitTryApplyInst(TryApplyInst *AI) {
     if (!AI->getOrigCalleeType()->isCalleeConsumed())
       return rewriteApplyCallee(AI, TTTFI->getOperand()).getInstruction();
   }
+
   // (apply (witness_method)) -> propagate information about
   // a concrete type from init_existential_addr or init_existential_ref.
   if (auto *WMI = dyn_cast<WitnessMethodInst>(AI->getCallee())) {
-    propagateConcreteTypeOfInitExistential(AI, WMI);
+    if (!propagateConcreteTypeOfInitExistential(AI, WMI)) {
+      if (auto *WitnessMethod = dyn_cast<WitnessMethodInst>(AI->getCallee())) {
+        // Propagate concrete type from ProtocolConformanceAnalysis.
+        propagateSoleConformingType(AI, WitnessMethod);
+      }
+    }
     return nullptr;
   }
 

--- a/lib/SILOptimizer/Utils/Existential.cpp
+++ b/lib/SILOptimizer/Utils/Existential.cpp
@@ -10,6 +10,8 @@
 //
 //===----------------------------------------------------------------------===//
 
+#include "swift/AST/Module.h"
+#include "swift/AST/ProtocolConformance.h"
 #include "swift/SILOptimizer/Utils/Existential.h"
 #include "swift/SILOptimizer/Utils/CFG.h"
 #include "swift/SIL/InstructionUtils.h"
@@ -307,5 +309,78 @@ ConcreteExistentialInfo::ConcreteExistentialInfo(Operand &openedUse) {
           InitExistential->getTypeDependentOperands()[0].get());
     }
   }
+  assert(isValid());
+}
+
+/// Initialize a ConcreteExistentialInfo based on the already computed concrete
+/// type and protocol declaration. It determines the OpenedArchetypeDef
+/// and SubstituionMap for the ArgOperand argument.
+/// We need the OpenedArchetypeDef to be able to cast it to the concrete type.
+/// findInitExistential helps us determine this OpenedArchetypeDef. For cases
+/// where OpenedArchetypeDef can not be found from findInitExistential (because
+/// there was no InitExistential), then we do
+/// extra effort in trying to find an OpenedArchetypeDef for AllocStackInst
+/// using getAddressOfStackInit.
+ConcreteExistentialInfo::ConcreteExistentialInfo(Operand &ArgOperand,
+                                                 CanType ConcreteTy,
+                                                 ProtocolDecl *Protocol)
+    : ConcreteExistentialInfo(ArgOperand) {
+
+  // If we found an InitExistential, assert that ConcreteType we determined is
+  // same as ConcreteTy argument.
+  if (InitExistential) {
+    assert(ConcreteType == ConcreteTy);
+    assert(isValid());
+    return;
+  }
+
+  ConcreteType = ConcreteTy;
+  OpenedArchetypeDef = ArgOperand.get();
+
+  // If findInitExistential call from the other constructor did not update the
+  // OpenedArchetypeDef (because it did not find an InitExistential) and that
+  // the original Arg is an alloc_stack instruction, then we determine the
+  // OpenedArchetypeDef using getAddressOfStackInit. Please keep in mind that an
+  // alloc_stack can be an argument to apply (which could have no associated
+  // InitExistential), thus we need to determine the OpenedArchetypeDef for it.
+  // This is the extra effort.
+  SILInstruction *User = ArgOperand.getUser();
+  SILModule &M = User->getModule();
+  if (auto *ASI = dyn_cast<AllocStackInst>(OpenedArchetypeDef)) {
+    bool copied = false;
+    if (SILValue Src = getAddressOfStackInit(ASI, User, copied)) {
+      OpenedArchetypeDef = Src;
+    }
+    isCopied = copied;
+  }
+
+  // Bail, if we did not find an opened existential.
+  if (!(isa<OpenExistentialRefInst>(OpenedArchetypeDef) ||
+        isa<OpenExistentialAddrInst>(OpenedArchetypeDef)))
+    return;
+
+  // We have the open_existential; we still need the conformance.
+  auto ConformanceRef =
+      M.getSwiftModule()->lookupConformance(ConcreteType, Protocol);
+  if (!ConformanceRef)
+    return;
+
+  // Assert that the conformance is complete.
+  auto *ConcreteConformance = ConformanceRef.getValue().getConcrete();
+  assert(ConcreteConformance->isComplete());
+
+  /// Determine the ExistentialConformances and SubstitutionMap.
+  ExistentialType = Protocol->getDeclaredType()->getCanonicalType();
+  auto ExistentialSig = M.getASTContext().getExistentialSignature(
+      ExistentialType, M.getSwiftModule());
+  ExistentialSubs = SubstitutionMap::get(
+      ExistentialSig, {ConcreteType},
+      llvm::makeArrayRef(ProtocolConformanceRef(ConcreteConformance)));
+
+  /// Determine the OpenedArchetype.
+  OpenedArchetype =
+      OpenedArchetypeDef->getType().castTo<ArchetypeType>();
+
+  /// Check validity.
   assert(isValid());
 }

--- a/test/SILOptimizer/sil_combine_concrete_existential.swift
+++ b/test/SILOptimizer/sil_combine_concrete_existential.swift
@@ -86,6 +86,12 @@ struct S: PPP {
   }
 }
 
+struct SS: PPP {
+  func returnsOptionalIndirect() -> SS? {
+    return self
+  }
+}
+
 // The first apply has been devirtualized and inlined. The second remains unspecialized.
 // CHECK-LABEL: sil @$S32sil_combine_concrete_existential37testWitnessReturnOptionalIndirectSelfyyF : $@convention(thin) () -> () {
 // CHECK: switch_enum_addr %{{.*}} : $*Optional<@opened("{{.*}}") PPP>, case #Optional.some!enumelt.1: bb{{.*}}, case #Optional.none!enumelt: bb{{.*}}

--- a/test/SILOptimizer/sil_combine_protocol_conf.swift
+++ b/test/SILOptimizer/sil_combine_protocol_conf.swift
@@ -1,0 +1,375 @@
+// RUN: %target-swift-frontend %s -O -wmo -emit-sil -Xllvm -sil-disable-pass=DeadFunctionElimination | %FileCheck %s
+
+// case 1: class protocol -- should optimize
+internal protocol SomeProtocol : class {
+  func foo(x:SomeProtocol)  -> Int
+  func foo_internal()  -> Int
+}
+internal class SomeClass: SomeProtocol {
+  func foo_internal() ->Int {
+    return 10
+  }
+  func foo(x:SomeProtocol) -> Int {
+    return x.foo_internal()
+  }
+}
+
+// case 2: non-class protocol -- should optimize
+internal protocol SomeNonClassProtocol {
+  func bar(x:SomeNonClassProtocol)  -> Int
+  func bar_internal()  -> Int
+}
+internal class SomeNonClass: SomeNonClassProtocol {
+  func bar_internal() -> Int {
+    return 20
+  }
+  func bar(x:SomeNonClassProtocol) -> Int {
+    return x.bar_internal()
+  }
+}
+
+// case 3: class conforming to protocol has a derived class -- should not optimize
+internal protocol DerivedProtocol {
+  func foo()  -> Int
+}
+internal class SomeDerivedClass: DerivedProtocol {
+  func foo() -> Int {
+    return 20
+  }
+}
+internal class SomeDerivedClassDerived: SomeDerivedClass {
+}
+
+// case 3: public protocol -- should not optimize
+public protocol PublicProtocol {
+  func foo()  -> Int
+}
+internal class SomePublicClass: PublicProtocol {
+  func foo() -> Int {
+    return 20
+  }
+}
+
+// case 4: Chain of protocols P1->P2->C -- optimize
+internal protocol MultiProtocolChain {
+  func foo()  -> Int
+}
+internal protocol RefinedMultiProtocolChain : MultiProtocolChain {
+  func bar()  -> Int
+}
+internal class SomeMultiClass: RefinedMultiProtocolChain {
+  func foo() -> Int {
+    return 20
+  }
+  func bar() -> Int {
+    return 30
+  }
+}
+
+// case 5: Generic conforming type -- should not  optimize
+internal protocol GenericProtocol {
+  func foo() -> Int 
+}
+
+internal class GenericClass<T> : GenericProtocol {
+  var items = [T]()
+  func foo() -> Int { 
+    return items.count 
+  }
+}
+
+
+// case 6: two classes conforming to protocol
+internal protocol MultipleConformanceProtocol {
+  func foo()  -> Int
+}
+internal class Klass1: MultipleConformanceProtocol {
+  func foo() -> Int {
+    return 20
+  }
+}
+internal class Klass2: MultipleConformanceProtocol {
+  func foo() -> Int {
+    return 30
+  }
+}
+
+
+internal class Other {
+   let x:SomeProtocol
+   let y:SomeNonClassProtocol
+   let z:DerivedProtocol
+   let p:PublicProtocol
+   let q:MultiProtocolChain
+   let r:GenericProtocol
+   let s:MultipleConformanceProtocol
+   init(x:SomeProtocol, y:SomeNonClassProtocol, z:DerivedProtocol, p:PublicProtocol, q:MultiProtocolChain, r:GenericProtocol, s:MultipleConformanceProtocol) {
+     self.x = x;
+     self.y = y;
+     self.z = z;
+     self.p = p;
+     self.q = q;
+     self.r = r;
+     self.s = s;
+   }
+
+// CHECK-LABEL: sil hidden [noinline] @$S25sil_combine_protocol_conf5OtherC11doWorkClassSiyF : $@convention(method) (@guaranteed Other) -> Int {
+// CHECK: bb0
+// CHECK: debug_value
+// CHECK: integer_literal
+// CHECK: [[R1:%.*]] = ref_element_addr %0 : $Other, #Other.z
+// CHECK: [[O1:%.*]] = open_existential_addr immutable_access [[R1]] : $*DerivedProtocol to $*@opened("{{.*}}") DerivedProtocol
+// CHECK: [[W1:%.*]] = witness_method $@opened("{{.*}}") DerivedProtocol, #DerivedProtocol.foo!1 : <Self where Self : DerivedProtocol> (Self) -> () -> Int, [[O1]] : $*@opened("{{.*}}") DerivedProtocol : $@convention(witness_method: DerivedProtocol) <τ_0_0 where τ_0_0 : DerivedProtocol> (@in_guaranteed τ_0_0) -> Int
+// CHECK: apply [[W1]]<@opened("{{.*}}") DerivedProtocol>([[O1]]) : $@convention(witness_method: DerivedProtocol) <τ_0_0 where τ_0_0 : DerivedProtocol> (@in_guaranteed τ_0_0) -> Int
+// CHECK: struct_extract
+// CHECK: integer_literal
+// CHECK: builtin
+// CHECK: tuple_extract 
+// CHECK: tuple_extract 
+// CHECK: cond_fail
+// CHECK: [[R2:%.*]] = ref_element_addr %0 : $Other, #Other.p
+// CHECK: [[O2:%.*]] = open_existential_addr immutable_access [[R2]] : $*PublicProtocol to $*@opened("{{.*}}") PublicProtocol
+// CHECK: [[W2:%.*]] = witness_method $@opened("{{.*}}") PublicProtocol, #PublicProtocol.foo!1 : <Self where Self : PublicProtocol> (Self) -> () -> Int, [[O2]] : $*@opened("{{.*}}") PublicProtocol : $@convention(witness_method: PublicProtocol) <τ_0_0 where τ_0_0 : PublicProtocol> (@in_guaranteed τ_0_0) -> Int
+// CHECK: apply [[W2]]<@opened("{{.*}}") PublicProtocol>([[O2]]) : $@convention(witness_method: PublicProtocol) <τ_0_0 where τ_0_0 : PublicProtocol> (@in_guaranteed τ_0_0) -> Int
+// CHECK: struct_extract
+// CHECK: builtin 
+// CHECK: tuple_extract
+// CHECK: tuple_extract
+// CHECK: cond_fail
+// CHECK: integer_literal
+// CHECK: builtin 
+// CHECK: tuple_extract
+// CHECK: tuple_extract
+// CHECK: cond_fail
+// CHECK: [[R3:%.*]] = ref_element_addr %0 : $Other, #Other.r
+// CHECK: [[O3:%.*]] = open_existential_addr immutable_access [[R3]] : $*GenericProtocol to $*@opened("{{.*}}") GenericProtocol
+// CHECK: [[W3:%.*]] = witness_method $@opened("{{.*}}") GenericProtocol, #GenericProtocol.foo!1 : <Self where Self : GenericProtocol> (Self) -> () -> Int, [[O3]] : $*@opened("{{.*}}") GenericProtocol : $@convention(witness_method: GenericProtocol) <τ_0_0 where τ_0_0 : GenericProtocol> (@in_guaranteed τ_0_0) -> Int
+// CHECK: apply [[W3]]<@opened("{{.*}}") GenericProtocol>([[O3]]) : $@convention(witness_method: GenericProtocol) <τ_0_0 where τ_0_0 : GenericProtocol> (@in_guaranteed τ_0_0) -> Int
+// CHECK: struct_extract
+// CHECK: builtin 
+// CHECK: tuple_extract
+// CHECK: tuple_extract
+// CHECK: cond_fail
+// CHECK: [[R4:%.*]] = ref_element_addr %0 : $Other, #Other.s
+// CHECK: [[O4:%.*]] = open_existential_addr immutable_access %36 : $*MultipleConformanceProtocol to $*@opened("{{.*}}") MultipleConformanceProtocol
+// CHECK: [[W4:%.*]] = witness_method $@opened("{{.*}}") MultipleConformanceProtocol, #MultipleConformanceProtocol.foo!1 : <Self where Self : MultipleConformanceProtocol> (Self) -> () -> Int, %37 : $*@opened("{{.*}}") MultipleConformanceProtocol : $@convention(witness_method: MultipleConformanceProtocol) <τ_0_0 where τ_0_0 : MultipleConformanceProtocol> (@in_guaranteed τ_0_0) -> Int
+// CHECK: apply [[W4]]<@opened("{{.*}}") MultipleConformanceProtocol>(%37) : $@convention(witness_method: MultipleConformanceProtocol) <τ_0_0 where τ_0_0 : MultipleConformanceProtocol> (@in_guaranteed τ_0_0) -> Int
+// CHECK: struct_extract
+// CHECK: builtin
+// CHECK: tuple_extract
+// CHECK: tuple_extract
+// CHECK: cond_fail
+// CHECK: struct 
+// CHECK: return
+// CHECK: } // end sil function '$S25sil_combine_protocol_conf5OtherC11doWorkClassSiyF'
+   @inline(never) func doWorkClass () ->Int {
+      return self.x.foo(x:self.x) // optimize
+             + self.y.bar(x:self.y) // optimize
+             + self.z.foo() // do not optimize
+             + self.p.foo() // do not optimize
+             + self.q.foo() // optimize
+             + self.r.foo() // do not optimize
+             + self.s.foo() // do not optimize
+   }
+}
+
+// case 1: struct -- optimize
+internal protocol PropProtocol {
+  var val: Int { get set }
+}
+internal struct PropClass: PropProtocol {
+  var val: Int
+  init(val: Int) {
+    self.val = val
+  }
+}
+
+// case 2: generic struct -- do not optimize
+internal protocol GenericPropProtocol {
+  var val: Int { get set }
+}
+internal struct GenericPropClass<T>: GenericPropProtocol {
+  var val: Int
+  init(val: Int) {
+    self.val = val
+  }
+}
+
+// case 3: nested struct -- optimize
+internal protocol NestedPropProtocol {
+  var val: Int { get }
+}
+struct Outer {
+  struct Inner : NestedPropProtocol {
+    var val: Int
+    init(val: Int) {
+      self.val = val
+    }
+  }
+}
+
+// case 4: generic nested struct -- do not optimize
+internal protocol GenericNestedPropProtocol {
+  var val: Int { get }
+}
+struct GenericOuter<T> {
+  struct GenericInner : GenericNestedPropProtocol {
+    var val: Int
+    init(val: Int) {
+      self.val = val
+    }
+  }
+}
+
+internal class OtherClass {
+  var arg1: PropProtocol
+  var arg2: GenericPropProtocol
+  var arg3: NestedPropProtocol
+  var arg4: GenericNestedPropProtocol
+
+  init(arg1:PropProtocol, arg2:GenericPropProtocol, arg3: NestedPropProtocol, arg4: GenericNestedPropProtocol) {
+     self.arg1 = arg1
+     self.arg2 = arg2
+     self.arg3 = arg3
+     self.arg4 = arg4
+  }
+
+// CHECK-LABEL: sil hidden [noinline] @$S25sil_combine_protocol_conf10OtherClassC12doWorkStructSiyF : $@convention(method) (@guaranteed OtherClass) -> Int {
+// CHECK: bb0
+// CHECK: debug_value
+// CHECK: [[A1:%.*]] = alloc_stack $PropProtocol
+// CHECK: [[R1:%.*]] = ref_element_addr %0 : $OtherClass, #OtherClass.arg1
+// CHECK: copy_addr [[R1]] to [initialization] [[A1]] : $*PropProtocol
+// CHECK: [[O1:%.*]] = open_existential_addr immutable_access [[A1]] : $*PropProtocol to $*@opened("{{.*}}") PropProtocol
+// CHECK: [[U1:%.*]] = unchecked_addr_cast [[O1]] : $*@opened("{{.*}}") PropProtocol to $*PropClass 
+// CHECK: [[S1:%.*]] = struct_element_addr [[U1]] : $*PropClass, #PropClass.val
+// CHECK: [[S11:%.*]] = struct_element_addr [[S1]] : $*Int, #Int._value
+// CHECK: load [[S11]] 
+// CHECK: destroy_addr [[A1]] : $*PropProtocol
+// CHECK: [[A2:%.*]] = alloc_stack $GenericPropProtocol 
+// CHECK: [[R2:%.*]] = ref_element_addr %0 : $OtherClass, #OtherClass.arg2
+// CHECK: copy_addr  [[R2]] to [initialization] [[A2]] : $*GenericPropProtocol
+// CHECK: [[O2:%.*]] = open_existential_addr immutable_access [[A2]] : $*GenericPropProtocol to $*@opened("{{.*}}") GenericPropProtocol
+// CHECK: [[W2:%.*]] = witness_method $@opened("{{.*}}") GenericPropProtocol, #GenericPropProtocol.val!getter.1 : <Self where Self : GenericPropProtocol> (Self) -> () -> Int, [[O2]] : $*@opened("{{.*}}") GenericPropProtocol : $@convention(witness_method: GenericPropProtocol) <τ_0_0 where τ_0_0 : GenericPropProtocol> (@in_guaranteed τ_0_0) -> Int
+// CHECK: apply [[W2]]<@opened("{{.*}}") GenericPropProtocol>([[O2]]) : $@convention(witness_method: GenericPropProtocol) <τ_0_0 where τ_0_0 : GenericPropProtocol> (@in_guaranteed τ_0_0) -> Int
+// CHECK: destroy_addr [[A2]] : $*GenericPropProtocol
+// CHECK: struct_extract
+// CHECK: integer_literal
+// CHECK: builtin
+// CHECK: tuple_extract
+// CHECK: tuple_extract
+// CHECK: cond_fail
+// CHECK: dealloc_stack [[A2]] : $*GenericPropProtocol
+// CHECK: dealloc_stack [[A1]] : $*PropProtocol
+// CHECK: [[A4:%.*]] = alloc_stack $NestedPropProtocol
+// CHECK: [[R4:%.*]] = ref_element_addr %0 : $OtherClass, #OtherClass.arg3
+// CHECK: copy_addr [[R4]] to [initialization] [[A4]] : $*NestedPropProtocol
+// CHECK: [[O4:%.*]] = open_existential_addr immutable_access [[A4]] : $*NestedPropProtocol to $*@opened("{{.*}}") NestedPropProtocol
+// CHECK: [[AA:%.*]] = alloc_stack $@opened("{{.*}}") NestedPropProtocol
+// CHECK: copy_addr [[O4]] to [initialization] [[AA]] : $*@opened("{{.*}}") NestedPropProtocol
+// CHECK: destroy_addr [[A4]] : $*NestedPropProtocol
+// CHECK: [[U4:%.*]] = unchecked_addr_cast [[AA]] : $*@opened("{{.*}}") NestedPropProtocol to $*Outer.Inner
+// CHECK: [[S4:%.*]] = struct_element_addr [[U4]] : $*Outer.Inner, #Outer.Inner.val
+// CHECK: [[S41:%.*]] = struct_element_addr [[S4]] : $*Int, #Int._value
+// CHECK: load [[S41]]
+// CHECK: builtin
+// CHECK: tuple_extract
+// CHECK: tuple_extract
+// CHECK: cond_fail
+// CHECK: dealloc_stack [[A4]] : $*NestedPropProtocol
+// CHECK: [[A5:%.*]] = alloc_stack $GenericNestedPropProtocol
+// CHECK: [[R5:%.*]] = ref_element_addr %0 : $OtherClass, #OtherClass.arg4
+// CHECK:  copy_addr [[R5]] to [initialization] [[A5]] : $*GenericNestedPropProtocol
+// CHECK: [[O5:%.*]] = open_existential_addr immutable_access [[A5]] : $*GenericNestedPropProtocol to $*@opened("{{.*}}") GenericNestedPropProtocol
+// CHECK: [[A6:%.*]] = alloc_stack $@opened("{{.*}}") GenericNestedPropProtocol
+// CHECK: copy_addr [[O5]] to [initialization] [[A6]] : $*@opened("{{.*}}") GenericNestedPropProtocol
+// CHECK: destroy_addr [[A5]] : $*GenericNestedPropProtocol
+// CHECK: [[W5:%.*]] = witness_method $@opened("{{.*}}") GenericNestedPropProtocol, #GenericNestedPropProtocol.val!getter.1 : <Self where Self : GenericNestedPropProtocol> (Self) -> () -> Int, [[O5:%.*]] : $*@opened("{{.*}}") GenericNestedPropProtocol : $@convention(witness_method: GenericNestedPropProtocol) <τ_0_0 where τ_0_0 : GenericNestedPropProtocol> (@in_guaranteed τ_0_0) -> Int 
+// CHECK: apply [[W5]]<@opened("{{.*}}") GenericNestedPropProtocol>([[A6]]) : $@convention(witness_method: GenericNestedPropProtocol) <τ_0_0 where τ_0_0 : GenericNestedPropProtocol> (@in_guaranteed τ_0_0) -> Int
+// CHECK: destroy_addr [[A6]] : $*@opened("{{.*}}") GenericNestedPropProtocol
+// CHECK: struct_extract
+// CHECK: builtin
+// CHECK: tuple_extract
+// CHECK: tuple_extract
+// CHECK: cond_fail
+// CHECK: struct
+// CHECK: dealloc_stack [[A6]] : $*@opened("{{.*}}") GenericNestedPropProtocol
+// CHECK: dealloc_stack [[A5]] : $*GenericNestedPropProtocol
+// CHECK: return
+// CHECK: } // end sil function '$S25sil_combine_protocol_conf10OtherClassC12doWorkStructSiyF'
+  @inline(never) func doWorkStruct () -> Int{
+    return self.arg1.val  // optimize
+           + self.arg2.val  // do not optimize
+           + self.arg3.val  // optimize
+           + self.arg4.val  // do not optimize
+  }
+}
+
+// case 1: enum -- optimize
+internal protocol AProtocol {
+  var val: Int { get }
+}
+internal enum AnEnum : AProtocol {
+    case avalue
+    var val: Int {
+        switch self {
+        case .avalue:
+            return 10
+        }
+    }
+}
+
+// case 2: generic enum -- do not optimize
+internal protocol AGenericProtocol {
+  var val: Int { get }
+}
+internal enum AGenericEnum<T> : AGenericProtocol {
+    case avalue
+    var val: Int {
+        switch self {
+        case .avalue:
+            return 10
+        }
+    }
+}
+
+internal class OtherKlass {
+  var arg1: AProtocol
+  var arg2: AGenericProtocol
+
+  init(arg1:AProtocol, arg2:AGenericProtocol) {
+     self.arg1 = arg1
+     self.arg2 = arg2
+  }
+
+// CHECK-LABEL: sil hidden [noinline] @$S25sil_combine_protocol_conf10OtherKlassC10doWorkEnumSiyF : $@convention(method) (@guaranteed OtherKlass) -> Int {
+// CHECK: bb0
+// CHECK: debug_value
+// CHECK: integer_literal
+// CHECK: [[A1:%.*]] = alloc_stack $AGenericProtocol
+// CHECK: [[R1:%.*]] = ref_element_addr %0 : $OtherKlass, #OtherKlass.arg2
+// CHECK: copy_addr [[R1]] to [initialization] [[A1]] : $*AGenericProtocol 
+// CHECK: [[O1:%.*]] = open_existential_addr immutable_access [[A1]] : $*AGenericProtocol to $*@opened("{{.*}}") AGenericProtocol
+// CHECK: [[A2:%.*]] = alloc_stack $@opened("{{.*}}") AGenericProtocol
+// CHECK: copy_addr [[O1]] to [initialization] [[A2]] : $*@opened("{{.*}}") AGenericProtocol
+// CHECK: destroy_addr [[A1]] : $*AGenericProtocol
+// CHECK: [[W1:%.*]] = witness_method $@opened("{{.*}}") AGenericProtocol, #AGenericProtocol.val!getter.1 : <Self where Self : AGenericProtocol> (Self) -> () -> Int, [[O1]] : $*@opened("{{.*}}") AGenericProtocol : $@convention(witness_method: AGenericProtocol) <τ_0_0 where τ_0_0 : AGenericProtocol> (@in_guaranteed τ_0_0) -> Int
+// CHECK: apply [[W1]]<@opened("{{.*}}") AGenericProtocol>([[A2]]) : $@convention(witness_method: AGenericProtocol) <τ_0_0 where τ_0_0 : AGenericProtocol> (@in_guaranteed τ_0_0) -> Int
+// CHECK: destroy_addr [[A2]] : $*@opened("{{.*}}") AGenericProtocol
+// CHECK: struct_extract
+// CHECK: integer_literal
+// CHECK: builtin
+// CHECK: tuple_extract
+// CHECK: tuple_extract
+// CHECK: cond_fail
+// CHECK: struct
+// CHECK: dealloc_stack [[A2]] : $*@opened("{{.*}}") AGenericProtocol
+// CHECK: dealloc_stack [[A1]] : $*AGenericProtocol
+// CHECK: return
+// CHECK: } // end sil function '$S25sil_combine_protocol_conf10OtherKlassC10doWorkEnumSiyF'
+  @inline(never) func doWorkEnum() -> Int {
+    return self.arg1.val  // optimize
+           + self.arg2.val // do not optimize
+  }
+}
+

